snuba/state/cache/redis/backend.py

```diff
@@ -0,0 +1,220 @@
+        wait_queue_key = self.__build_key(key, "tasks", "wait")
+
+        # The task identity (a Redis bytestring) is used to store a unique
+        # identifier for a single task evaluation and notification cycle. Only
+        # one task should be associated with a cache key at any time.
+        task_ident_key = self.__build_key(key, "tasks")
+
+        # The notify queue (a Redis list) is used to unblock clients that are
+        # waiting for the task to complete. **This implementation requires that
+        # the number of clients waiting for responses via the notify queue is
+        # no greater than the number of clients in the wait queue** (minus one
+        # client, who is doing the work and not waiting.) If there are more
+        # clients waiting for notifications than the members of the wait queue
+        # (for any reason), some set of clients will never be notified. To be
+        # safe and ensure that each client only waits for notifications from
+        # tasks where it was also a member of the wait queue, the notify queue
+        # includes the unique task identity as part of it's key.
+        def build_notify_queue_key(task_ident: str) -> str:
+            return self.__build_key("key", "tasks", f"notify/{task_ident}")
+
+        # At this point, we have all of the information we need to figure out
+        # if the key exists, and if it doesn't, if we should start working or
+        # wait for a different client to finish the work. We have to pass the
+        # task creation parameters -- the timeout (execution deadline) and a
+        # new task identity just in case we are the first in line.
+        result = self.__script_get(
+            [result_key, wait_queue_key, task_ident_key], [timeout, uuid.uuid1().hex]
+        )
+
+        if result[0] == RESULT_VALUE:
+            # If we got a cache hit, this is easy -- we just return it.
+            logger.debug("Immediately returning result from cache hit.")
+            return self.__codec.decode(result[1])
+        elif result[0] == RESULT_EXECUTE:
+            # If we were the first in line, we need to execute the function.
+            # We'll also get back the task identity to use for sending
+            # notifications and approximately how long we have to run the
+            # function. (In practice, these should be the same values as what
+            # we provided earlier.)
+            task_ident = result[1].decode("utf-8")
+            task_timeout = int(result[2])
+            logger.debug(
+                "Executing task (%r) with %s second timeout...",
+                task_ident,
+                task_timeout,
+            )
+
+            argv = [task_ident, 60]
+            try:
+                # The task is run in a thread pool so that we can return
+                # control to the caller once the timeout is reached.
+                value = self.__executor.submit(function).result(task_timeout)
```

Comment:
Note that this will cause any exceptions raised by the function _after_ the timeout not to be logged, even though they might contain useful information. Should we hook up a callback to log them?

Reply:
No strong opinion.  It could be useful get the actual error, but it might be enough that we get the timeout exception. (After that point, aren't they all most likely timeouts anyway?)

Reply:
Going to punt on this for now, I think that if we decide we need this it's easy enough to add later.